### ISSUE TITLE
Making dataset

### ISSUE THREAD

**Issue by peanut1101** [ming024/FastSpeech2#166]
How to create dataset？

**Comment by ruslantau** [comment]
**First of all**, you need to collect audio data in wav/mp3 format and then use [this](https://github.com/ruslantau/media-annotator) or analog for annotation your audios. 
**2 step** – export your annotations as `csv` file and `.zip` of wavs(unzip this).
**3 step** – create configs for you data and run `python3 prepare_align.py config/LJSpeech/preprocess.yaml`.
This create `raw data dir` with `.wav`s and `.lab` files, which contain text from corresponding wav.
**4 step** – install [MFA](https://montreal-forced-aligner.readthedocs.io/en/latest/) or another aligner and:

- create pronunciation dictionary for your data [using mfa](https://montreal-forced-aligner.readthedocs.io/en/latest/user_guide/workflows/dictionary_generating.html#mfa-g2p) or another tool 

- [train mfa aligner](https://montreal-forced-aligner.readthedocs.io/en/latest/user_guide/workflows/train_acoustic_model.html#mfa-train) 
example: `mfa train raw_data/your_dataset/ lexicon/your_dataset.txt out_dir/aligner_model.zip preprocessed_data/your_dataset/TextGrid`

**5 step** – run `python3 preprocess.py config/LJSpeech/preprocess.yaml`. It create `preprocessed_data` dir with data for training.
**6 step** – run model [training](https://github.com/ming024/FastSpeech2#training-1)

**Comment by alessandropec** [comment]
Thanks a lot for your contribution, @ruslantau i have a question, in order to train i need als speakers.json (this is ok is just the mapping to id of each speaker in the dataset) stats.json, how to compute this? moreover i need the new list with transcription (val.txt and train.txt) how to generate those, i already have the lists for train and val but i have the text not codified in the phonemas extracted by the mfa, how can i do it?

**Comment by ryamldess** [comment]
I'm trying to do this as well and still find it confusing after reading @ruslantau 's steps above. I want to use the LibreTTS dataset, but looking inside, each speaker has a folder that contains an XX.book.tsv file and an XX.trans.tsv file. Are these files generated by the tool linked above? Or does it happen in the MFA steps? If it is in the MFA steps, then why do we have to perform this step on the pre-existing LibreTTS model if we want to train on that, as it appears that this has already been completed and is part of the dataset to begin with? I'm trying to fine-tune to perform voice cloning, but I cannot get started because I do not even know how to format the data in order to tack it onto LibreTTS for training.

**Comment by ryamldess** [comment]
I've found the overall documentation and responses to this question here and elsewhere inadequate, so I chatted with ChatGPT to generate a comprehensive walkthrough. I have edited it slightly to make it more suitable for documentation rather than a chat. Due to the chat nature of ChatGPT, it turns out a bit like an FAQ:

I am attempting to fine-tune a FastSpeech2 model with custom data, a series of voice recordings in .WAV format. I am attempting to convert my small test dataset of WAV files into the TextGrid format FastSpeech2 requires using the preferred tool for this, called MFA. Is that something you think that you can help me with?

Before we get started, it's worth noting that the MFA tool is designed to align text with speech, and it requires both a text transcript and a corresponding audio file to work. If you don't have a text transcript for your WAV files, you'll need to create one before proceeding with the alignment process.

Assuming you have a text transcript for each of your WAV files, here are the general steps you'll need to follow to use MFA for alignment:

1. Install the MFA tool: The first step is to install the MFA tool on your computer. You can find installation instructions on the MFA GitHub page (https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner).

2. Prepare your data: Before you can use MFA to align your data, you'll need to prepare it by organizing your audio and text files in a specific directory structure. You'll need to create a "corpus" directory that contains subdirectories for each speaker in your dataset, and each speaker directory should contain two subdirectories: one for audio files and one for text files. Your audio files should be in WAV format, and your text files should be in plain text format with one transcript per file.

3. Prepare your pronunciation dictionary: MFA uses a pronunciation dictionary to map words to phonemes. You'll need to create a pronunciation dictionary that includes all the words in your text transcripts and their corresponding phonemes. You can create a dictionary manually, or you can use a tool like G2P-seq2seq to automatically generate phonetic transcriptions for your words.

4. Run the alignment process: Once you've prepared your data and pronunciation dictionary, you can run the alignment process using the MFA tool. You'll need to specify the location of your corpus directory, the location of your pronunciation dictionary, and some other options related to the alignment process. The output of the alignment process will be TextGrid files that contain the phoneme alignments for each of your audio files.

5. Convert TextGrid to the required format: The final step is to convert the TextGrid files to the required format for use with FastSpeech2. According to the FastSpeech2 documentation, the required format is a tab-separated text file with the following columns: "start time", "end time", and "phoneme". You can use a tool like Praat or a script to extract the required information from the TextGrid files and create the tab-separated text files.

The corpus directory needs to be named exactly that, i.e. the string "corpus", is that correct? What do I name the folders containing the audio and text files? Can I name those two folders anything, or do I have to name them something specific?

Yes, the corpus directory needs to be named exactly "corpus" (all lowercase) for the MFA tool to recognize it as a corpus directory.

As for the folders containing the audio and text files, you can name them anything you like, as long as they are consistent with the directory structure that MFA expects. The convention is to use the speaker's name as the name of the directory that contains their audio and text files. For example, if you have two speakers named John and Jane, you could create a directory structure like this:

arduino
Copy code
corpus/
├── john/
│   ├── audio/
│   │   ├── john_001.wav
│   │   ├── john_002.wav
│   │   └── ...
│   ├── text/
│   │   ├── john_001.txt
│   │   ├── john_002.txt
│   │   └── ...
│   └── lexicon.txt
├── jane/
│   ├── audio/
│   │   ├── jane_001.wav
│   │   ├── jane_002.wav
│   │   └── ...
│   ├── text/
│   │   ├── jane_001.txt
│   │   ├── jane_002.txt
│   │   └── ...
│   └── lexicon.txt
└── ...

In this example, each speaker has their own directory under the corpus directory, and each speaker directory contains an audio directory and a text directory. The audio directory contains the WAV files for that speaker, and the text directory contains the corresponding text transcripts. The lexicon.txt file is the pronunciation dictionary for that speaker.

So to summarize, you can name the folders containing the audio and text files anything you like, as long as they are organized in a directory structure that is consistent with the convention described above.

MFA uses file and folder recursion to automatically locate the audio and text files in the corpus directory and its subdirectories. This means that you can organize your audio and text files into subdirectories based on any scheme that makes sense for your project, and MFA will still be able to find them as long as they are located somewhere within the corpus directory.

For example, if you had a corpus directory with the following structure:

arduino
Copy code
corpus/
├── speaker1/
│   ├── audio/
│   │   ├── recording1.wav
│   │   ├── recording2.wav
│   │   └── ...
│   ├── text/
│   │   ├── recording1.txt
│   │   ├── recording2.txt
│   │   └── ...
│   └── lexicon.txt
├── speaker2/
│   ├── audio/
│   │   ├── recording1.wav
│   │   ├── recording2.wav
│   │   └── ...
│   ├── text/
│   │   ├── recording1.txt
│   │   ├── recording2.txt
│   │   └── ...
│   └── lexicon.txt
└── ...

MFA would be able to find all of the audio and text files for speaker1 and speaker2, even though they are located in subdirectories of the corpus directory.

It's important to note that the audio and text files must have the same basename (i.e., the same filename without the extension) in order for MFA to match them up correctly. For example, if you have an audio file named "recording1.wav" and a text file named "recording1.txt", MFA will assume that they correspond to each other and process them accordingly.

If I wanted to make a manual pronunciation dictionary, how would I even go about that?

Creating a manual pronunciation dictionary can be a bit time-consuming, but it can be a useful way to ensure that the TTS model generates accurate pronunciations for words that may not be covered in a pre-existing dictionary.

Here are some general steps to create a manual pronunciation dictionary:

Determine the words that need to be added to the dictionary. This may involve reviewing the text transcripts of your audio files and identifying any words that are not recognized correctly by your automatic speech recognition (ASR) system, or words that are specific to your domain or language and are not covered by a pre-existing dictionary.

Research the correct pronunciation of each word. This may involve looking up the word in a dictionary, listening to audio recordings of native speakers pronouncing the word, or consulting with a language expert.

Decide on the phonetic transcription of each word. Phonetic transcriptions represent the individual sounds that make up a word, and can be used by TTS models to generate accurate pronunciations. There are a few different systems for phonetic transcription, but the most common is the International Phonetic Alphabet (IPA). You can use an online tool or a reference book to learn the IPA symbols for each sound in your language.

Add the words and their phonetic transcriptions to a text file in the format required by your TTS system. For example, in the case of the FastSpeech2 implementation you linked to, the format for the pronunciation dictionary is a plain text file with one word and its phonetic transcription per line, separated by a tab character. Here's an example entry for the word "hello" in the CMU Pronouncing Dictionary format:

Copy code
HELLO	HH AH L OW
Save the pronunciation dictionary file in the appropriate directory in your corpus directory, as described in step 2.

By adding custom pronunciations for specific words to your pronunciation dictionary, you can improve the accuracy and naturalness of your TTS system's output.

If I wanted to automate it with G2P-seq2seq, is that fairly straightforward as well, or is there a sequence of disambiguated steps you can share about that tool as well?

G2P-seq2seq is a tool that uses a sequence-to-sequence neural network model to learn the mapping between orthographic (written) words and their phonetic transcriptions. It can be a useful tool for automatically generating phonetic transcriptions for words that are not covered by a pre-existing dictionary.

Here are the general steps for using G2P-seq2seq:

Install G2P-seq2seq on your system. You can find installation instructions in the G2P-seq2seq GitHub repository.

Prepare a training corpus. This should be a text file containing a list of words and their correct phonetic transcriptions. You can use a pre-existing pronunciation dictionary as a starting point, or create your own using the manual process described in my previous answer.

Train the G2P-seq2seq model on your training corpus. This involves specifying the hyperparameters of the model and running the training script. You can find detailed instructions on how to do this in the G2P-seq2seq documentation.

Use the trained model to generate phonetic transcriptions for new words. This involves running the prediction script and providing it with a text file containing a list of words you want to transcribe. The script will output a file containing the predicted phonetic transcriptions.

Add the new words and their phonetic transcriptions to your TTS pronunciation dictionary file.

Note that training a G2P-seq2seq model can be computationally intensive and may require a significant amount of training data to achieve good performance. However, once the model is trained, generating phonetic transcriptions for new words is a relatively quick and straightforward process.

You mention I need to "You'll need to specify the location of your corpus directory, the location of your pronunciation dictionary, and some other options related to the alignment process." Where and how do I specify all of these parameters?

To specify the parameters for running MFA, you'll need to use the command line interface. Here's an example of the basic command for running MFA, with explanations of each parameter:

mfa_align /path/to/corpus /path/to/dictionary /path/to/output/directory

/path/to/corpus: This is the path to the directory containing your audio files and their corresponding text files in TextGrid format.
/path/to/dictionary: This is the path to your pronunciation dictionary file.
/path/to/output/directory: This is the path to the directory where you want the aligned output files to be saved.
In addition to these three required parameters, there are many other optional parameters that you can use to customize the alignment process. Here are a few examples:

--num_jobs: This parameter controls the number of parallel jobs that will be used to align the data. By default, MFA will use all available CPU cores.
--config_file: This parameter specifies the path to a configuration file that can be used to customize various aspects of the alignment process, such as the acoustic model used for alignment.
--temp_directory: This parameter specifies the path to a temporary directory where MFA can store intermediate files during the alignment process.
To see a full list of available options and their descriptions, you can run the following command:

mfa_align --help

Does Praat do everything to the create the tab-separated files for me? These are the files in .tab format, correct? What are the exact steps?

Yes, Praat can be used to create the tab-separated files needed for training FastSpeech2. These files are in .tab format and contain information about the timing and duration of each phoneme in the input audio.

Here are the steps to create the tab-separated files using Praat:

1. Open Praat and load the audio file you want to process.
2. Use the "Read from file..." command in the "TextGrid" menu to load the TextGrid file that you generated in step 4 of the MFA process. This should open the audio file in a new window with the TextGrid overlay displayed.
3. Use the "Extract intervals and tier names" command in the "TextGrid" menu to extract the timing information from the TextGrid file. This will create a new table that displays the start and end times of each interval, as well as the associated phoneme label.
4. Use the "Save as..." command in the "Table" menu to save the table as a tab-separated text file. Make sure to save it with a .tab extension.
5. Once you've created the tab-separated file for your audio file, you can use it to train your FastSpeech2 model. Just repeat this process for each audio file in your corpus, and generate a corresponding tab-separated file for each one.

**Comment by phamkhactu** [comment]
Hi @alessandropec,

Have you find a ways to make ``` stas.json``` ?

**Comment by djanibekov** [comment]
@phamkhactu read this part https://github.com/ming024/FastSpeech2/blob/d4e79eb52e8b01d24703b2dfc0385544092958f3/preprocessor/preprocessor.py#L116
